src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -1036,6 +1089,69 @@ private static bool IsPrimaryKeySort(SearchOptions searchOptions)
             return searchOptions.Sort.All(s => s.searchParameterInfo.Name is SearchParameterNames.ResourceType or SearchParameterNames.LastUpdated);
         }
 
+        private bool IsSortValueNeeded(SearchOptions context)
+        {
+            if (context.Sort.Count == 0)
+            {
+                return false;
+            }
+
+            if (IsPrimaryKeySort(context))
+            {
+                return false;
+            }
+
+            foreach (var searchParamTableExpression in _rootExpression.SearchParamTableExpressions)
+            {
+                if (searchParamTableExpression.Kind == SearchParamTableExpressionKind.Sort ||
+                    searchParamTableExpression.Kind == SearchParamTableExpressionKind.SortWithFilter)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static SortContext GetSortRelatedDetails(SearchOptions context)
+        {
+            SortContext sortContext = new SortContext();
+            SearchParameterInfo searchParamInfo = default;
+            if (context.Sort?.Count > 0)
+            {
+                (searchParamInfo, sortContext.SortOrder) = context.Sort[0];
+            }
+
+            sortContext.ContinuationToken = ContinuationToken.FromString(context.ContinuationToken);
+
+            if (searchParamInfo.Type == ValueSets.SearchParamType.Date)
+            {
+                sortContext.SortColumnName = VLatest.DateTimeSearchParam.StartDateTime;
+            }
+            else if (searchParamInfo.Type == ValueSets.SearchParamType.String)
+            {
+                sortContext.SortColumnName = VLatest.StringSearchParam.Text;
+            }
```

Comment:
switch statement? Assuming this will keep growing

Reply:
Refactored with a switch statement.